### PR TITLE
feat(auth): GetUserByProviderID()

### PR DESCRIPTION
**Pull request by rsgowman**

### CODE REVIEW

**Comment by rsgowman** [comment]
Note the diffbase. I'll merge+retarget to dev once the bulk_get stuff lands.

File: auth/user_mgt.go [hunk 1]
```diff
@@ -544,6 +544,41 @@ func (c *baseClient) GetUserByPhoneNumber(ctx context.Context, phone string) (*U
 	})
 }
 
+// GetUserByProviderUid gets the user data for the user corresponding to a given provider id.
```
Comment (hiranya911) [inline auth/user_mgt.go]: GetUserByProviderID

File: auth/user_mgt_test.go [hunk 2]
```diff
@@ -140,22 +140,118 @@ func TestGetUserByPhoneNumber(t *testing.T) {
...
+		t.Errorf("GetUserByProviderID() Req = %v; want = %v", got, want)
+	}
+
+	wantPath := "/projects/mock-project-id/accounts:lookup"
+	if s.Req[0].RequestURI != wantPath {
+		t.Errorf("GetUserByProviderID() URL = %q; want = %q", s.Req[0].RequestURI, wantPath)
+	}
+}
+
+func TestGetEmailUserByProviderID(t *testing.T) {
```
Comment (hiranya911) [inline auth/user_mgt_test.go]: You can possibly combine these last 3 test cases into one table-driven test case.

```
var cases []struct{
  providerID string
  providerUID string
  want string
} {
  {
    "email",
    "user@example.com",
   `{"email": ["user@example.com"]}`,
  },
  // Other cases
}
```

File: auth/user_mgt_test.go [hunk 3]
```diff
@@ -140,22 +140,118 @@ func TestGetUserByPhoneNumber(t *testing.T) {
 	}
 }
 
+func TestGetUserByProviderIDNotFound(t *testing.T) {
```
Comment (hiranya911) [inline auth/user_mgt_test.go]: Also test for empty providerID, empty providerUID etc

File: auth/user_mgt.go (hunk 1 above)
Comment (rsgowman) [reply auth/user_mgt.go]: Done.

File: auth/user_mgt_test.go (hunk 2 above)
Comment (rsgowman) [reply auth/user_mgt_test.go]: Done.

File: auth/user_mgt_test.go (hunk 3 above)
Comment (rsgowman) [reply auth/user_mgt_test.go]: See below; I added them to the TestInvalidGetUser() test. (Arguably, that could be 4 or 5 different tests, but they are at least very tightly related.)

File: auth/user_mgt_test.go [hunk 4]
```diff
@@ -152,72 +152,49 @@ func TestGetUserByProviderIDNotFound(t *testing.T) {
...
-}
 
-func TestGetEmailUserByProviderID(t *testing.T) {
-	// The resulting user isn't parsed, so it just needs to exist (even if it's empty).
-	mockUsers := []byte(`{ "users": [{}] }`)
-	s := echoServer(mockUsers, t)
-	defer s.Close()
+	for _, tc := range cases {
+		t.Run(tc.providerID+":"+tc.providerUID, func(t *testing.T) {
+			// The resulting user isn't parsed, so it just needs to exist (even if it's empty).
```
[the remark this answers is not in the document]
Comment (rsgowman) [reply auth/user_mgt_test.go]: Done.

File: .github/scripts/publish_post_check.sh [hunk 5]
```diff
@@ -0,0 +1,105 @@
+
```
Comment (hiranya911) [inline .github/scripts/publish_post_check.sh:1]: This file no longer exists in source. Can be dropped from this PR.

File: .github/scripts/publish_preflight_check.sh [hunk 6]
```diff
@@ -92,6 +92,8 @@ fi
 echo_info "Tag v${RELEASE_VERSION} does not exist."
 
 
+<<<<<<< HEAD
```
Comment (hiranya911) [inline .github/scripts/publish_preflight_check.sh:95]: Please revert.

File: .github/workflows/publish.yml [hunk 7]
```diff
@@ -0,0 +1,64 @@
+# Copyright 2020 Google Inc.
```
Comment (hiranya911) [inline .github/workflows/publish.yml:1]: Not needed.

File: .github/workflows/stage.yml [hunk 8]
```diff
@@ -0,0 +1,74 @@
+# Copyright 2020 Google Inc.
```
Comment (hiranya911) [inline .github/workflows/stage.yml:1]: Not needed.

File: auth/user_mgt_test.go [hunk 9]
```diff
@@ -141,22 +141,96 @@ func TestGetUserByPhoneNumber(t *testing.T) {
...
 }
 
+func TestGetUserByProviderIDNotFound(t *testing.T) {
+	mockUsers := []byte(`{ "users": [] }`)
+	s := echoServer(mockUsers, t)
+	defer s.Close()
+
+	userRecord, err := s.Client.GetUserByProviderID(context.Background(), "google.com", "google_uid1")
+	want := "cannot find user from providerID: { google.com, google_uid1 }"
+	if userRecord != nil || err == nil || err.Error() != want {
```
Comment (hiranya911) [inline auth/user_mgt_test.go:151]: Might want to also check for `IsUserNotFoundError(err)` here

**Review by hiranya911 — CHANGES_REQUESTED** [review]
Implementation and tests look pretty good. Lets remove the redundant CI config bits, and we should be good to go.

File: .github/scripts/publish_post_check.sh (hunk 5 above)
Comment (rsgowman) [reply .github/scripts/publish_post_check.sh]: Ah, that's what caused this bad merge. (I mean, I should've caught it anyways, but this is making more sense now.)

File: .github/scripts/publish_preflight_check.sh (hunk 6 above)
Comment (rsgowman) [reply .github/scripts/publish_preflight_check.sh]: Done.

File: .github/workflows/publish.yml (hunk 7 above)
Comment (rsgowman) [reply .github/workflows/publish.yml]: Done.

File: .github/workflows/stage.yml (hunk 8 above)
Comment (rsgowman) [reply .github/workflows/stage.yml]: Done.

File: auth/user_mgt_test.go (hunk 9 above)
Comment (rsgowman) [reply auth/user_mgt_test.go]: Done.

**Review by hiranya911 — APPROVED** [review]
Thanks! LGTM 👍

File: auth/user_mgt.go [hunk 10]
```diff
@@ -498,6 +499,47 @@ func (c *baseClient) GetUserByPhoneNumber(ctx context.Context, phone string) (*U
 	})
 }
 
+// GetUserByProviderID gets the user data for the user corresponding to a given provider id.
```
Comment (egilmorez) [inline auth/user_mgt.go:502]: Unless this is a literal, capped "ID" would probably be preferred.

File: auth/user_mgt.go [hunk 11]
```diff
@@ -498,6 +499,47 @@ func (c *baseClient) GetUserByPhoneNumber(ctx context.Context, phone string) (*U
...
+//
+// See
+// [Retrieve user data](https://firebase.google.com/docs/auth/admin/manage-users#retrieve_user_data)
+// for code samples and detailed documentation.
+//
+// `providerID` indicates the provider, e.g. 'google.com' for the Google provider.
```
Comment (egilmorez) [inline auth/user_mgt.go:508]: Suggest "such as" instead of "e.g."

My sense is that if we want to say "For example," it should come after a hard stop like a period or semicolon after "provider."  Internal style guide: https://developers.google.com/style/abbreviations#dont-use

File: auth/user_mgt.go [hunk 12]
```diff
@@ -498,6 +499,47 @@ func (c *baseClient) GetUserByPhoneNumber(ctx context.Context, phone string) (*U
...
+// `providerUID` is the user identifier for the given provider.
+func (c *baseClient) GetUserByProviderID(ctx context.Context, providerID string, providerUID string) (*UserRecord, error) {
+	// Although we don't really advertise it, we want to also handle non-federated
+	// idps with this call. So if we detect one of them, we'll reroute this
```
Comment (egilmorez) [inline auth/user_mgt.go:512]: Should this be IDPs, with the acronym in caps?

Comment (egilmorez) [reply auth/user_mgt.go:512]: Actually, this may not be a public-facing comment I guess :)

**Review by egilmorez — COMMENTED** [review]
Just some minor nits.  See what you think, and I'm happy to approve in any case.  Thanks!

File: auth/user_mgt.go (hunk 10 above)
Comment (rsgowman) [reply auth/user_mgt.go]: Done.

File: auth/user_mgt.go (hunk 11 above)
Comment (rsgowman) [reply auth/user_mgt.go]: Done.

File: auth/user_mgt.go (hunk 12 above)
Comment (rsgowman) [reply auth/user_mgt.go]: Not a public facing comment; but still a valid point, so done anyways.

**Review by egilmorez — APPROVED** [review]
LG, thanks!
